controllers/geotags/dynamic.go

```diff
@@ -0,0 +1,102 @@
+package geotags
+
+/*
+Copyright 2022 The k8gb Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Generated by GoLic, for more details see: https://github.com/AbsaOSS/golic
+*/
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/k8gb-io/k8gb/controllers/depresolver"
+	"github.com/k8gb-io/k8gb/controllers/utils"
+	"github.com/miekg/dns"
+	"go.uber.org/multierr"
+)
+
+// Dynamic reads geotags from parent DNS
+type Dynamic struct {
+	config *depresolver.Config
+}
+
+func NewDynamic(config *depresolver.Config) *Dynamic {
+	return &Dynamic{
+		config: config,
+	}
+}
+
+func (p *Dynamic) GetExternalClusterNSNamesByHostname(host string) (map[string]string, error) {
+	z := getZone(p.config, host)
+	if z != nil {
+		return p.getExternalClusterNSNamesByHostname()
+	}
+	return map[string]string{}, nil
+}
+
+func (p *Dynamic) getExternalClusterNSNamesByHostname() (map[string]string, error) {
+	var tags []string
+	var err error
+	var resultErr error
+	nsNames := make(map[string]string)
+	for _, edge := range p.config.ParentZoneDNSServers {
+		for _, d := range p.config.DelegationZones {
+			tags, err = p.getExternalTags(edge, d.LoadBalancedZone)
+			if err != nil {
+				resultErr = multierr.Append(resultErr, fmt.Errorf("error getting tags for zone %s: %w", d.ParentZone, err))
+				continue
+			}
+			for _, tag := range tags {
+				nsNames[tag] = depresolver.GetNsName(tag, d.LoadBalancedZone, d.ParentZone)
+			}
+		}
+	}
+	return nsNames, resultErr
+}
+
+func (p *Dynamic) getExternalTags(edge utils.DNSServer, zone string) ([]string, error) {
+	const prefix = "gslb-ns-"
+	var extTags []string
+	m := new(dns.Msg)
+	m.SetQuestion(zone+".", dns.TypeNS)
+	m.RecursionDesired = false // Equivalent to dig +norec
+	c := new(dns.Client)
+	r, _, err := c.Exchange(m, edge.String())
+	if err != nil {
+		return extTags, err
+	}
+
+	for _, ans := range r.Ns {
+		ns, ok := ans.(*dns.NS)
+		if !ok {
+			continue
+		}
+		// Extract tag from NS name pattern: gslb-ns-tag-zone
+		parts := strings.Split(ns.Ns, prefix)
+		// filter non k8gb ns records
+		if len(parts) != 2 {
+			continue
+		}
+		// parts[0] == prefix; parts[1] == <region>-test-cloud.example.com
+		tag := strings.Split(parts[1], "-")[0]
+		if tag != p.config.ClusterGeoTag {
+			extTags = append(extTags, tag)
+		}
+	}
+	sort.Strings(extTags)
```

Comment:
IMO we should exclude `own` tag

Reply:
sure, we have it: https://github.com/k8gb-io/k8gb/pull/1914/files#diff-42234b232ea30a245f58cd1db685055d2b547cb01090dd25e4a8389befd048a3R96-R98

Reply:
I see